APPS-1378 dxcint async tests

JanProvaznik opened the pull request:
In the end there is no asynchronous code, only concurrent. Since the tests are independent, I simply gave a thread to each one (preemptive multithreading) and they can use DXJob.wait_for_done without interfering with each other. Performance should be OK unless there are >1000 tests.

Gvaihir committed:
dxcint
(fix) Terraform.py - no chdir in threads.

JanProvaznik committed:
(fix) test failures due to duplicate builds

JanProvaznik committed:
dxcint

(feat) GHA for testing dxcint - initial

JanProvaznik committed:
dxcint

(fix) dxcint tests

JanProvaznik committed:
Merge branch 'APPS-1389-dxCint-test-fails' into APPS-1366-dxcint-test-GHA

JanProvaznik committed:
dxcint

(fix) dxcint -locked in RegisteredTest compilation

Gvaihir committed:
Merge branch 'develop' into APPS-1366-dxcint-test-GHA

Gvaihir committed:
APPS-1366 dxcint test GHA
(fix) RegisteredTest.py - removed `-locked`. See comments to #380.
(fix) mock_1_input.json - applied Jan's suggestions

JanProvaznik committed:
dxcint

(fix) dxcint.yml - simplify installation with poetry
(fix) test_RegisteredTest.py - update input name

JanProvaznik committed:
minor

JanProvaznik committed:
minor

JanProvaznik committed:
minor

JanProvaznik committed:
dxcint

(feat) cli.py - concurrent dxcint tests
(feat) Messenger.py, RegisteredTest.py, testclasses/ExpectedFailure.py - concurrent dxcint tests
(feat) test_Messenger.py, conftest.py - test concurrent behaviour
(fix) Dependency.py - apply linter suggestions
(fix) TestDiscovery.py - ignore this class in pytest
(fix) Terrafrom.py - remove pointless locking

JanProvaznik committed:
Merge branch 'develop' into APPS-1378-dxcint-async-tests

Gvaihir commented on dxcint/dxcint/TestDiscovery.py:
```diff
@@ -19,6 +19,7 @@ class TestDiscoveryError(Exception):
 
 
 class TestDiscovery(object):
+    __test__ = False
```
good catch

Gvaihir commented on dxcint/dxcint/RegisteredTest.py:
```diff
@@ -132,8 +134,8 @@ def _run_executable(self) -> str:
-        exec_desc = self._run_executable_inner().describe()
-        return exec_desc.get("id")
+        execution_desc = self._run_executable_inner().describe()
+        return execution_desc.get("id")
```
ok, more readable

Gvaihir commented on dxcint/dxcint/RegisteredTest.py:
```diff
@@ -32,8 +32,8 @@ def __init__(
         self._test_name = test_name
         self._language = rm_prefix(os.path.basename(src_file), test_name).strip(".")
         self._messenger = None
```
self._create_messenger()

JanProvaznik replied:
No, the messenger is and should be created only after the test is ran, which is not the same as when the class is created.

Gvaihir replied:
right, there's a dependency on job ID

Gvaihir commented on dxcint/dxcint/cli.py:
```diff
@@ -69,8 +71,21 @@ def integration(
-        passed = test.test_result
+
+    # simply give every test a thread, no need for cooperative multitasking,
+    # tests are independent, they will print correctly progress to logger
```
```suggestion
    # tests are independent, they will correctly print the progress to the logger
```

Gvaihir commented on dxcint/dxcint/cli.py:
```diff
@@ -108,7 +123,8 @@ def add(
     """
-    test_context = Context(project="dxCompiler_playground", repo_root=dxc_repository_root)
+    test_context = Context(project="dxCompiler_playground",
+                           repo_root=dxc_repository_root)
```
Pls follow [black](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#parentheses). I'll create a ticket for you to add pre-commit action for black and flake8 for the python code.

Gvaihir commented on dxcint/dxcint/cli.py:
```diff
@@ -69,8 +71,21 @@ def integration(
+             for registered_test in registered_tests}
+        results: List[bool] = [f.result()
+                   for f in futures.as_completed(future_to_execute_tests)]
+    failures = results.count(False)
```
`num_failures` or `failure_counter`

Gvaihir commented on dxcint/tests/conftest.py:
```diff
@@ -5,6 +5,7 @@
 from dxcint.Context import Context
 from dxcint.Dependency import DependencyFactory
 from dxcint.Terraform import Terraform
+from .test_RegisteredTest import build_executable_wdl, registered_test_wdl
```
please no relative imports. Also conftest should not import anything to potentially avoid cyclic imports. Just move those fixtures to conftest

Gvaihir commented on dxcint/dxcint/cli.py:
```diff
@@ -69,8 +71,21 @@ def integration(
+            {executor.submit(registered_test.get_test_result) 
+             for registered_test in registered_tests}
+        results: List[bool] = [f.result()
+                   for f in futures.as_completed(future_to_execute_tests)]
```
pls follow black

Gvaihir commented on dxcint/dxcint/testclasses/ExpectedFailure.py:
```diff
@@ -12,13 +11,9 @@ def __init__(self, src_file: str, category: str, test_name: str, context: Contex
         return execution.describe().get("id")
 
-    def _validate(self) -> Dict:
+    async def _validate(self) -> Dict:
```
why making it a coroutine? You are not using `await`. Or is it a carry-over form other implementation attempts?

JanProvaznik replied:
thanks for spotting it, forgot it there, deleted in the next commit

Gvaihir commented on dxcint/dxcint/Messenger.py:
```diff
@@ -1,60 +1,70 @@
+        self._test_name = test_name
+        self._variant = variant if variant else ""
+        self._state: Union[State, str] = State.NOT_DONE
+        self._job: Union[dxpy.DXAnalysis, dxpy.DXJob] = None
```
Redundant because next line

Gvaihir commented on dxcint/dxcint/Messenger.py:
```diff
@@ -1,60 +1,70 @@
+        """
+        logging.info(
+            f"Waiting for completion of the test `{self._test_name}` with input variant {self._variant}")
+        test_name_with_variant = ".".join(
```
f-string better?

Gvaihir commented on dxcint/dxcint/Messenger.py:
```diff
@@ -1,60 +1,70 @@
+        if 'analysis-' in job_id:
+            self._job = dxpy.DXAnalysis(job_id)
+        else:
+            self._job = dxpy.DXJob(job_id)
```
check `dxpy.bindings.dxdataobject_functions.get_handler` - maybe no need for a conditional

Gvaihir commented on dxcint/dxcint/Messenger.py:
```diff
@@ -1,60 +1,70 @@
+            self._state = State.FAIL
+        return self.state
+
+    def query_state(self, extended: bool = False) -> Union[State, str]:
```
good, handy

Gvaihir commented on dxcint/dxcint/Messenger.py:
```diff
@@ -1,60 +1,70 @@
+        return self.state
+
+    def query_state(self, extended: bool = False) -> Union[State, str]:
+        """Queries the state of the job and returns it in the form of a State enum."""
```
If method has args and returns stuff please include those in the docstring

Gvaihir commented on dxcint/tests/test_Messenger.py:
```diff
@@ -1,9 +1,45 @@
+        history2 = h2.result()
+
+    assert results == [False, False]
+    assert State.NOT_DONE in history1
```
Can you test for the order at which those statuses appear to confirm concurrency? For example - create a dict with both histories and update it with the status changes. Then check that test 1 did not wait for test 2 to finish

JanProvaznik replied:
Added a simple check for that the jobs share some time in NOT_DONE state but beyond that the workers are quite unpredictable, so one waiting and one being done is not impossible even if they are submitted at the same time.

Gvaihir reviewed (changes requested):
Great job! Requested changes. Will add a ticket for code linter for this package and precommits.
EDIT: APPS-1397

JanProvaznik committed:
dxcint

(test) test_Messenger - test that the executions overlap
(refactor) test_Messenger, conftest.py
(style) cli - grammar, naming
(refactor) Messenger - simpler typing
(fix) ExpectedFailure - remove async
(style) all files in dxcint - apply black formatter

Gvaihir reviewed (approved):
Looks good!
